requests/models.py

```diff
@@ -796,7 +800,23 @@ def iter_lines(self, chunk_size=ITER_CHUNK_SIZE, decode_unicode=None, delimiter=
             if delimiter:
                 lines = chunk.split(delimiter)
             else:
+                # Python splitlines() supports the universal newline (PEP 278).
+                # That means, '\r', '\n', and '\r\n' are all treated as end of
+                # line. If the last chunk ends with '\r', and the current chunk
+                # starts with '\n', they should be merged and treated as only
+                # *one* new line separator '\r\n' by splitlines().
+                # This rule only applies when splitlines() is used.
+                if last_chunk_ends_with_cr and chunk.startswith(line_feed):
+                    # The last chunk ends with '\r', so the '\n' at chunk[0]
+                    # is just the second half of a '\r\n' pair rather than a
+                    # new line break. Just skip it.
+                    chunk = chunk[1:]
+                    # it's possible that after stripping the '\n' then chunk becomes empty
+                    if not chunk:
+                        continue
                 lines = chunk.splitlines()
+                # check if the current chunk ends with '\r'
+                last_chunk_ends_with_cr = chunk.endswith(carriage_return)
```

Comment:
Oooh, just noticed a subtle misbehaviour. Imagine we get the following chunks:

```python
[b'this is a string\r', b'\n', b'\nso is this']
```

This will not do the right thing: we'll strip *two* newline characters, rather than emit an empty line. That is, we'll emit: `[b'this is a string', b'so is this']` instead of `[b'this is a string', b'', b'so is this']`, which is what we *should* emit.

This is because a chunk that is *just* a `\n` causes an early continue, above, without resetting the value of the `last_chunk_ends_with_cr` boolean. Essentially it gets completely ignored. That's not right.

It'd be good to add a test case (or modify an existing one) to include that case, and then fix it up by hoisting this flag up before the `if not chunk: continue` clause.